### ISSUE TITLE
How to install v1.12-dev.5 on OSX?

### ISSUE THREAD

**Issue by svetlyak40wt** [roswell/roswell#405]
Hi!

Clozure CL 1.11 is broken on OSX Mojave. Is there a way to install 1.12 using Roswell?

**Comment by snmsts** [comment]
no way so far.　

Are there reason clozure guys don't prepare release binary?
https://github.com/roswell/daily_maintenance/blob/master/ccl-bin/Lakefile
I'm prepared for release somehow..

**Comment by svetlyak40wt** [comment]
There are few issues left till the release. I found the roadmap:

https://github.com/Clozure/ccl/projects/3

However, I don't know when developers are planning to get those issues done :(

**Comment by svetlyak40wt** [comment]
Probably, there is a way to tell Roswell: "Hey, here is a binary I've downloaded, consider it is a ccl-bin/1.12"?

**Comment by snmsts** [comment]
I'm not so curious now about v1.12 as release management. But somehow disappointed about there github releases usage. why there's  v1.11.7 tag without release? why they don't care about breakage for months. sorry I know you're not responsible about it. and I understand there lack of resources as well. just sad.


> Probably, there is a way to tell Roswell: "Hey, here is a binary I've downloaded, consider it is a ccl-bin/1.12"?

I wrote a patch to specify uri of the archive.

```
ros download.uri=https://example.com/archive.tar.gz install ccl-bin/1.12
```

It is easier fix for me than prepare new local copy installation
and I believe it is not hard for you to prepare proper archive on web.

I have no idea what is the best way to manage https://github.com/roswell/ccl_bin/releases If someone interest in releasing I would like to give access to the repo.

**Comment by svetlyak40wt** [comment]
Are there instructions how to prepare a tar archive for use like `download.uri=https://example.com/archive.tar.gz`?

**Comment by snmsts** [comment]
refer https://github.com/roswell/ccl_bin/releases

how about

```
ros download.uri=https://github.com/roswell/ccl_bin/releases/download/1.11.5/ccl-1.11.5-darwinx86.tar.gz install ccl-bin/1.12
```

**Comment by svetlyak40wt** [comment]
Oh, ok. This question was so stupid! 😔

**Comment by svetlyak40wt** [comment]
Well, I've tried to use `download.url` but it does not work.

```
ros download.url=https://github.com/Clozure/ccl/releases/download/v1.12-dev.5/darwinx86.
tar.gz install ccl-bin                                                                                             Checking version to install...
Installing ccl-bin/1.11.5...
Skip downloading https://github.com/roswell/ccl_bin/releases/download/1.11.5/ccl-1.11.5-darwinx86.tar.gz
specify download.force=t to download it again.

Extracting archive:/Users/art/.roswell/archives/ccl-1.11.5-darwinx86.tar.gz
```

and if I specify a version in fails with another error:

```
ros download.url=https://github.com/Clozure/ccl/releases/download/v1.12-dev.5/darwinx86.
tar.gz install ccl-bin/1.12                                                                                        Installing ccl-bin/1.12...
Downloading https://github.com/roswell/ccl_bin/releases/download/1.12/ccl-1.12-darwinx86.tar.gz
Downloading archive:https://github.com/roswell/ccl_bin/releases/download/1.12/ccl-1.12-darwinx86.tar.gz:NG        Unhandled SIMPLE-ERROR in thread #<SB-THREAD:THREAD "main thread" RUNNING
                                    {10004F84C3}>:                                                                   FOO BAR: args: (TARGET ccl-bin VERSION 1.12 VERSION-NOT-SPECIFIED NIL ARGV
                  NIL) condition: Subprocess with command "/usr/local/bin/ros roswell-internal-use download https://github.com/roswell/ccl_bin/releases/download/1.12/ccl-1.12-darwinx86.tar.gz /Users/art/.roswell/archives/ccl-1.12-
darwinx86.tar.gz"                                                                                                   exited with error code 2
```

Pay attention, that it tries to download not from the URL I've specified, but from the `https://github.com/roswell/ccl_bin/releases/download/1.12/ccl-1.12-darwinx86.tar.gz`

The roswell's version I'm using is `19.12.13.103(7b99c8b)`.

I've added some `print` statements into the `ROSWELL.INSTALL:INSTALL` and found that is is called with a single argument: `ccl-bin/1.12`.

@snmsts how does downoad.url should be passed into the `ROSWELL.UTIL:DOWNLOAD` function?

**Comment by snmsts** [comment]
```
ros -v download.uri=https://github.com/Clozure/ccl/releases/download/v1.12-dev.5/darwinx86.tar.gz install ccl-bin/1.12
```
this would work (until download)

try after
```
mv ~/.roswell ~/.roswellbak
```

**Comment by svetlyak40wt** [comment]
Yeah, this almost work except CCL's original archive format have a little bit different layout. I'll make a small patch to make it work, ok?

**Comment by snmsts** [comment]
No.
fix the archive please.

**Comment by svetlyak40wt** [comment]
By the way, difference between the `download.uri` and `download.url` is confusing, may be we can make it more user friendly by adding an automatic conversion `url -> uri` somewhere in the `ros-options`?

**Comment by svetlyak40wt** [comment]
> No.
Why? I can fix archive for myself, but it will be much simpler for other users of the Roswell, if it will accept official CCL urls.

**Comment by snmsts** [comment]
I can't stand they change archive layout time by time. they should fix there release archive.
If current archive is the fixed one I'll accept but I believe they don't think about it at all. I can't accept it.

**Comment by snmsts** [comment]
how about supported old archives? drop and make new archives for olds? or filled with ``if`` for versions? who maintain it?

**Comment by svetlyak40wt** [comment]
I'll support this patch in my fork untill the official 1.12 release. Probably this will will be helpful for other OSX and CCL users.

**Comment by svetlyak40wt** [comment]
Also, I've made this issue on CCL repository. Maybe they will change something in their release cycle:

https://github.com/Clozure/ccl/issues/269

**Comment by svetlyak40wt** [comment]
Here is the branch with a fix: https://github.com/svetlyak40wt/roswell/tree/accept-ccl-release-urls
and commit https://github.com/svetlyak40wt/roswell/commit/e1562eedaa2d690113222337891b8607318f0e1c in case if somebody will find this issue will want to install 1.12 before the official release.

**Comment by snmsts** [comment]
Thank you.
Let me know. I'll review your patch.

**Comment by snmsts** [comment]
oh already...

**Comment by svetlyak40wt** [comment]
@snmsts seems you was right. CCL downloaded from  is not fully functional and requires following the instruction from https://github.com/Clozure/ccl/releases/tag/v1.12-dev.5.

Without sources, for example `(require "COCOA")` outputs the following error:

```
? (require "COCOA")
> Error: Module COCOA was not provided by any function on *MODULE-PROVIDER-FUNCTIONS*.
> While executing: REQUIRE, in process listener(1).
> Type :POP to abort, :R for a list of available restarts.
> Type :? for other options.
1 >
```

Excuse me for bothering you. I have to build a separate archive :(

**Comment by snmsts** [comment]
Thank you very much for considering making release archives :).I'm thinking about support ``dev.5`` like version name. currently I haven't consider support such names.
